src/libproc_macro/lib.rs

```diff
@@ -212,12 +213,62 @@ impl Span {
         ::__internal::with_sess(|(_, mark)| Span(mark.expn_info().unwrap().call_site))
     }
 
+    /// The original source file into which this span points.
+    #[unstable(feature = "proc_macro", issue = "38356")]
+    pub fn source_file(&self) -> Option<&Path> {
+        // If this span's token was produced by a macro
+        if self.0.ctxt() != SyntaxContext::empty() {
+            return None
+        }
+
+        let loc = __internal::lookup_char_pos(self.0.lo());
+
+        Some(Path::new(&loc.file.name))
```

Comment:
@jseyfried That's right, this borrow is tied to `loc` instead of `self` so we can't return `&Path` here. Should I put it back to returning an opaque wrapper that can be dereferenced to `Path`, or just return a copy in `PathBuf`?

Reply:
I think returning a copy in a `PathBuf` is fine, but I think the return value should be a newtype anyway for future comparability.

Reply:
Well if we return a newtype then getting `&Path` from it is as easy as returning `PathBuf`. Is a `PathBuf` still preferable in case we change the internals such that it's not easy to get a borrow that lives long enough?

Reply:
So long as it's future compatible (i.e. just `impl Deref { type Target = Path; }`) it's up to you -- if internals change we can change later.